### PR TITLE
Change default type handling behaviour - Part 1

### PR DESCRIPTION
**Pull request by EmilyBourne**
In order to deprecate header functions we need to implement #297 to replace the .pyh files in `stdlib/internals`. This is a big change so it must be done in stages.
This PR moves the treatment of a variable header to the semantic stage. It introduces many of the classes and functions which will be necessary to do the same for function headers, but is less complex. This makes the review simpler.

Add a try/except clause round textx calls to report clean header syntax errors.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
/bot run coverage pylint pyccel_lint spelling docs

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:2094]: /bot accept
Blocked by #1527

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:2085]: /bot accept
Blocked by #1527

**Comment by yguclu** [comment]
Can you please explain the role of `TypeAnnotation`, `SyntacticTypeAnnotation` and `TypeAnnotation` during the parsing process?

**Comment by yguclu** [comment]
> Can you please explain the role of `TypeAnnotation`, `SyntacticTypeAnnotation` and `UnionTypeAnnotation` during the parsing process?

@EmilyBourne Thanks for the clear explanation you have added to the developers' documentation!

I would like to suggest a change in the names of these classes. As they are right now, I would expect `SyntacticTypeAnnotation` and `FunctionTypeAnnotation` to be subclasses of `TypeAnnotation`, which clearly is not the case. Hence I suggest to rename `TypeAnnotation` as `VariableTypeAnnotation`. This would also be in line with the Python documentation (see e.g. [PEP 593](https://peps.python.org/pep-0593/)).

**Comment by EmilyBourne** [comment]
> > Can you please explain the role of `TypeAnnotation`, `SyntacticTypeAnnotation` and `UnionTypeAnnotation` during the parsing process?
> 
> @EmilyBourne Thanks for the clear explanation you have added to the developers' documentation!
> 
> I would like to suggest a change in the names of these classes. As they are right now, I would expect `SyntacticTypeAnnotation` and `FunctionTypeAnnotation` to be subclasses of `TypeAnnotation`, which clearly is not the case. Hence I suggest to rename `TypeAnnotation` as `VariableTypeAnnotation`. This would also be in line with the Python documentation (see e.g. [PEP 593](https://peps.python.org/pep-0593/)).

Fixed

File: tests/epyccel/test_epyccel_headers.py
Comment (yguclu) [inline tests/epyccel/test_epyccel_headers.py:61]: Does this mean that a stack array cannot be declared through a header? Or do we get a similar error using a variable annotation?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1426]: Hmm maybe. Given we never used (or documented) variable headers previously we did not realise that we had syntax to pre-declare a variable so we did not consider this case when writing the code. Given that it is not mentioned in the docs and we have never had someone open an issue about this I am not sure it is worth opening an issue to explain the problem just so it can be referred to in the CHANGELOG. Unless you have an easy summary that you think would be worth adding?

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1869]: This would be the case if we did:
```python
#$ header variable var int|float
```

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1869]: `line[0]` is a `Variable` if `line` is a `VariableHeader`

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1869]: This can be removed when headers are deprecated

File: tests/epyccel/test_epyccel_headers.py
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_headers.py:61]: Yes this is correct. This is because the array is created (including its fixed size) at line 56. We don't have information about its shape here so this is not possible. In one of the next few PRs I will fix #1528 . This will allow us to translate:
```
array : 'int[:,:]' = np.array([[1,2],[3,4],[5,6]])
```
Here all the information is in the same line. So there is no issue.

That said, the usual usage for this is in a header file where the declaration won't be printed. We may want to think about extending the syntax to allow a declaration of:
```
#$ header variable array int[2,3]
```

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1870]: Added a comment to this effect: 575a52fe

File: pyccel/ast/type_annotations.py
Comment (EmilyBourne) [reply pyccel/ast/type_annotations.py]: Fixed bdf2271a

File: pyccel/ast/internals.py
Comment (EmilyBourne) [reply pyccel/ast/internals.py:332]: This is what I did first. It caused problems in PR #1523. But I can't remember exactly what problems.
It makes it harder to detect if the `AnnotatedPyccelSymbol` is not correctly visited as `_visit_PyccelSymbol` is called instead.
I think there may also be a few places where it is useful to differentiate between `PyccelSymbol` and `AnnotatedPyccelSymbol`. As is this is simple, but if `AnnotatedPyccelSymbol` inherits from `PyccelSymbol` then the condition becomes more complex:
```python
if isinstance(var, PyccelSymbol) and not isinstance(var, AnnotatedPyccelSymbol)
```

If you think this is important then I can go back to the version where `AnnotatedPyccelSymbol` is a subclass.

I think I may have also had a problem in the scope as the symbols are used as keys the first time they are encountered. When there is a variable header the first time they are encountered they are `AnnotatedPyccelSymbol`s but then we look up the variable with a `PyccelSymbol`. This may have been the problem that caused me to change strategy, but thinking about it it should be possible to correct it by implementing `__eq__` instead.

Comment (EmilyBourne) [reply pyccel/ast/internals.py:332]: An additional argument comes from error raising. Errors are easily raised from an `AnnotatedPyccelSymbol`. This object should therefore have a `fst` property so the line and symbol can be shown in the error message. The handling of this is already done in `Basic`, but it would need to be added if `AnnotatedPyccelSymbol` subclassed `PyccelSymbol` as `PyccelSymbol`s are reused and are therefore not easily described by a single `fst` pattern.

Comment (EmilyBourne) [reply pyccel/ast/internals.py:332]: If I implement the duplicate methods then the tests seem to pass so I leave it up to you to decide which strategy you think makes more sense

File: pyccel/parser/semantic.py
Comment (yguclu) [reply pyccel/parser/semantic.py:1870]: Great, thanks!

File: tests/epyccel/test_epyccel_headers.py
Comment (yguclu) [reply tests/epyccel/test_epyccel_headers.py:61]: I see, thanks for the explanation!

The header file (or stub file) will declare the type of an array variable including its rank, but I do not see the need to know its exact shape given that we do not provide it for function arguments.

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_headers.py]: You are right. Fixed def16c2b

File: pyccel/parser/semantic.py
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1432]: Fixed [d3be2d7](https://github.com/pyccel/pyccel/pull/1524/commits/d3be2d761f9f355bfc77f52ef6f1b4448b66f098)

File: pyccel/ast/internals.py
Comment (yguclu) [reply pyccel/ast/internals.py:332]: `PyccelSymbol` objects are immutable, but the `fst` property of `AnnotatedPyccelNode` needs to be set after creating an object of this class. Hence it makes sense that `AnnotatedPyccelNode` should subclass `Basic` as in the current implementation.

**Review by yguclu — APPROVED** [review]

**Pull request merged**
